### PR TITLE
feat(E6data): initial implemantation for E6data SQL Analytics platform

### PR DESCRIPTION
**Pull request by hackintoshrao**
## Description of changes
This pull request aims to integrate [E6data](https://docs.e6data.com/product-documentation/introduction-to-e6data), a distributed SQL analytics engine, with Ibis. E6data is also designed for high-performance analytics on large-scale data, and this integration will allow Ibis users to leverage E6data's capabilities seamlessly.

## Key changes and considerations:

1. E6data Backend Implementation:
   - A new E6data backend class was added, inheriting from SQLBackend.
   - Implemented connection handling, query execution, and result fetching specific to E6data.
   - Adapted the backend to handle E6data's unique features, such as catalog support and cluster management.

2. SQL Dialect Customization:
   - Created an E6data dialect class extending from MySQL, as E6data shares similarities with MySQL syntax.
   - Customized the Tokenizer to use double quotes for identifiers.
   - Modified the Generator to map certain data types (VARCHAR, CHAR, TEXT) to STRING, aligning with E6data's type system.
   - Custom TRANSFORMS for specific SQL functions like concat and length were added.

3. Compiler Modifications:
   - Updated E6DataCompiler to use the new E6data dialect and E6DataType for type mapping.
   - Retained existing rewrites, including a custom limit rewrite, to ensure compatibility with E6data's query execution model.

4. Connection String and Authentication:
   - Implemented support for E6data's connection string format, including catalog name, secure connection, auto-resume, and cluster UUID parameters.

5. Schema and Metadata Handling:
   - Adapted schema retrieval and table listing functions to work with E6data's multi-level hierarchy (catalog, database, table).

6. Testing:
I need some guidance on how to go about adding tests for the integration.

7. Documentation:
   Could you give me some pointers on how to add relevant documentation?

8. Dependencies:
   - Currently, the platform is not open for public access; existing users require authentication keys to use the analytics engine. I would like guidance on enabling the maintainers to test and provide credentials for automated tests once they are supported.
We're also working on a Mini-kube-based single-node testing infrastructure, which might make adding testing automation for the CI easier.

I'm new to the Ibis community, and this PR could be much better. I appreciate the time and guidance from maintainers in improving it further. Any comments are welcome; again, I appreciate your time and patience.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -87,6 +87,7 @@ shapely = { version = ">=2,<3", optional = true }
 # issues with versions <3.0.2
 snowflake-connector-python = { version = ">=3.0.2,<4,!=3.3.0b1", optional = true }
 trino = { version = ">=0.321,<1", optional = true }
+e6data-python-connector = { version = "2.2.0" }
```
Comment (cpcloud) [inline pyproject.toml:90]:
```suggestion
e6data-python-connector = { version = ">=2.2.0,<3", optional = true }
```

Unless there's a reason to pin to a single version, this should include a version range, and `optional = true` is necessary to ensure this is shipped as an extra, and not a required dependency to use any other Ibis backend(s).

File: ibis/backends/sql/dialects.py [hunk 2]
```diff
@@ -403,6 +403,7 @@ class Generator(Postgres.Generator):
         JSON_TYPE_REQUIRED_FOR_EXTRACTION = True
         SUPPORTS_UNLOGGED_TABLES = True
 
+
```
Comment (cpcloud) [inline ibis/backends/sql/dialects.py:406]:
```suggestion
```

Please revert this change.

File: ibis/backends/e6data/__init__.py [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+import re
+import warnings
+from functools import cached_property, partial
+from itertools import repeat
+from operator import itemgetter
+from typing import TYPE_CHECKING, Any
+from urllib.parse import parse_qs, urlparse
+
+import numpy as np
+import pymysql
```
Comment (cpcloud) [inline ibis/backends/e6data/__init__.py:15]: Are you using this somewhere? If not, pleas remove it as an e6data backend dependency and also remove this import.

File: ibis/backends/e6data/tests/conftest.py
Comment (cpcloud) [inline ibis/backends/e6data/tests/conftest.py:1]: This doesn't seem like the correct implementation for an e6data `conftest.py` given that it's not running in a container.

File: pyproject.toml [hunk 4]
```diff
@@ -152,6 +153,7 @@ dask = ["dask", "regex", "packaging"]
 datafusion = ["datafusion"]
 druid = ["pydruid"]
 duckdb = ["duckdb"]
+e6data = ["e6data-python-connector","pymysql"]
```
Comment (cpcloud) [inline pyproject.toml:156]:
```suggestion
e6data = ["e6data-python-connector"]
```

Is `pymysql` actually required? It doesn't seem to be used anywhere.

File: ibis/backends/e6data/compiler.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    > SELECT * FROM tbl LIMIT 95,18446744073709551615;
+    """
+    if _.n is None and _.offset is not None:
+        some_large_number = (1 << 64) - 1
+        return _.copy(n=some_large_number)
+    return _
+
+
+@public
+class E6DataCompiler(SQLGlotCompiler):
```
Comment (cpcloud) [inline ibis/backends/e6data/compiler.py:45]: Why not inherit from `MySQLCompiler`?

File: ibis/backends/e6data/__init__.py [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+from e6data_python_connector import Connection
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+
+    import pandas as pd
+    import polars as pl
+    import pyarrow as pa
+
+class Backend(SQLBackend, CanCreateDatabase):
```
Comment (cpcloud) [inline ibis/backends/e6data/__init__.py:38]: With all the copypasting it _might_ be worth considering inheriting from `MySQLBackend`.

File: ibis/backends/e6data/__init__.py [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+    name = "e6data"
+    compiler = E6DataCompiler()
+    supports_create_or_replace = False
+
+    @cached_property
+    def version(self):
+        matched = re.search(r"(\d+)\.(\d+)\.(\d+)", self.con.server_version)
+        return ".".join(matched.groups())
+    def _from_url(self, url: str, **kwargs):
```
Comment (cpcloud) [inline ibis/backends/e6data/__init__.py:47]: Please make sure to run your code through `just fmt`. There are likely other places where there are formatting inconsistencies.

**Review by cpcloud — CHANGES_REQUESTED** [review]
Thanks for the PR! Did a first pass, and will add a separate comment addressing some of your questions.

**Comment by cpcloud** [comment]
> 6. Testing:
> I need some guidance on how to go about adding tests for the integration.

The best place to start is to try and run `pytest -m e6data`, assuming you have a way to access it from your development host.

> 7. Documentation:
>    Could you give me some pointers on how to add relevant documentation?

Take a look at the individual backend docs pages in `docs/backends/`. Those are a good place to start with backend-specific docs.

> 8. Dependencies:
>    - Currently, the platform is not open for public access; existing users require authentication keys to use the analytics engine. I would like guidance on enabling the maintainers to test and provide credentials for automated tests once they are supported.
We're also working on a Mini-kube-based single-node testing infrastructure, which might make adding testing automation for the CI easier.

We'll need to get whatever credentials/auth information is needed to login into a GitHub Actions secret. Let's chat in a DM on Zulip about this.

File: ibis/backends/e6data/tests/conftest.py
Comment (hackintoshrao) [reply ibis/backends/e6data/tests/conftest.py:1]: Removed the tests until implemented correctly.

**Comment by hackintoshrao** [comment]
Hey @cpcloud ,

I've addressed most of your comments apart from the tests and documentation. Please let me know if these changes look good. I'll be working on writing the tests now.

> We'll need to get whatever credentials/auth information is needed to login into a GitHub Actions secret. Let's chat in a DM on Zulip about this.

I've Dm'ed you on Zulip, kindly need your assistance.

**Comment by cpcloud** [comment]
Please create a new PR if you're still interested in working on this.

**Pull request closed without merging**